File: interval-list-intersections/interval-list-intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        answer = []
        if not firstList:
            return answer
        if not secondList:
            return answer
        
        first = second = 0
        while first < len(firstList) and second < len(secondList):
            lo = max(firstList[first][0], secondList[second][0])
            hi = min(firstList[first][1], secondList[second][1])
            if lo <= hi:
                answer.append([lo, hi])
            if firstList[first][1] < secondList[second][1]:
                first += 1
            else:
                second += 1

        return answer

File: interval-list-intersections/interval-list-intersections.java
class Solution 
{
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) 
    {
        List<int[]> answer = new ArrayList<>();
        int f = 0, s = 0;
        
        while (f < firstList.length && s < secondList.length)
        {
            // Find range of intersecting interval
            int min = Math.max(firstList[f][0], secondList[s][0]);
            int max = Math.min(firstList[f][1], secondList[s][1]);
            
            // If range is acceptable, append
            if (min <= max)
                answer.add(new int[]{min, max});
            
            // Move the iterators forward
            if (firstList[f][1] < secondList[s][1]) f += 1;
            else s += 1;
        }
        
        return answer.toArray(new int[answer.size()][]);
    }
}